components/eamxx/src/share/field/field_manager.hpp

```diff
@@ -65,73 +69,80 @@ class FieldManager {
-
-  const FieldIdentifier& get_field_id (const std::string& name) const;
-  Field get_field (const std::string& name) const;
-  Field get_field (const identifier_type& id) const { return get_field(id.name()); }
-  Field& get_field (const std::string& name);
-  Field& get_field (const identifier_type& id) { return get_field(id.name()); }
-
-  FieldGroup get_field_group (const std::string& name) const;
-
-  repo_type::const_iterator begin () const { return m_fields.cbegin(); }
-  repo_type::const_iterator end   () const { return m_fields.cend();   }
-
-  // Set the time stamp of all fields
+  bool has_field (const std::string& field_name, const std::string& grid_name) const;
+  bool has_group (const std::string& name, const std::string& grid_name) const;
+
+  const FieldIdentifier& get_field_id (const std::string& name, const std::string& grid_name) const;
+  Field get_field (const std::string& name, const std::string& grid_name) const;
+  Field get_field (const identifier_type& id) const { return get_field(id.name(), id.get_grid_name()); }
+  Field& get_field (const std::string& name, const std::string& grid_name);
+  Field& get_field (const identifier_type& id) { return get_field(id.name(), id.get_grid_name()); }
+  FieldGroup get_field_group (const std::string& name, const std::string& grid_name) const;
+
+  const std::map<ci_string,std::shared_ptr<Field>>&
+  get_repo (const std::string& grid_name) const {
+    EKAT_REQUIRE_MSG(m_grids_mgr->has_grid(grid_name),
+      "Error! This field manager does not contain data on grid \""+grid_name+"\"\n");
+    return m_fields.at(grid_name);
+  }
+
+  const grids_mgr_type& get_grids_manager () const { return m_grids_mgr; }
+
+  // Set the time stamp of all fields on given grid
   // TODO: I think I want to remove this. We don't want to blanket-init
   //       the time stamp. IC reader can init the ts of IC fields, then
   //       let the different atm procs update ts when needed.
-  void init_fields_time_stamp (const util::TimeStamp& t0);
+  void init_fields_time_stamp (const util::TimeStamp& t0, const std::string& grid_name);
 
 protected:
 
   // These are allowed even if registration is ongoing
-  std::shared_ptr<Field> get_field_ptr(const std::string& name) const;
+  std::shared_ptr<Field> get_field_ptr(const std::string& name, const std::string& grid_name) const;
   std::shared_ptr<Field> get_field_ptr(const identifier_type& id) const;
 
   void pre_process_group_requests ();
 
   // The state of the repository
-  RepoState           m_repo_state;
+  RepoState m_repo_state;
 
   // The actual repo.
   repo_type           m_fields;
 
-  // When registering subfields, we might end up registering the subfield before
-  // the parent field. So at registration time, simply keep track of the subfields,
-  // and create them at registration_ends() time, after all other fields.
-  std::map<std::string,FieldRequest> m_subfield_requests;
```

Comment:
Indeed, we were not correctly handling subfield requests. In particular, we were never setting up subfields at the end of registration_ends.

Since nobody has complained, I gather the incomplete "feature" was never used. Hence, I would also purge the reference to this possibility in FieldRequest (the subview-related members and the subview-related constructor).

Reply:
`FieldAtHeight` diag appears to be using incomplete FIDs. I can still remove from FM if you want, but I guess for now if incomplete FIDs exist out there, maybe we might as well keep in FM (in case they catch on).

Reply:
Ah, good catch. Diagnostics are peculiar, in the sense that we do not register in a FM, but simply grab the field from a FM.

Then let's leave the feature in FieldRequest, but let's purge from FM (as you did). The question is: what happens if an incomplete request is used (after this PR)? Does FM throw? I would like it to throw...

Reply:
I added an assert that FR input to register_field() is not incomplete.